drv/stm32h7-spi-server-core/src/lib.rs

```diff
@@ -0,0 +1,767 @@
+            None,
+        )
+    }
+
+    pub fn exchange<
+        'b,
+        BufRead: BufReader<'b>,
+        Read: Into<BufRead> + 'b,
+        BufWrite: BufWriter<'b>,
+        Write: Into<BufWrite> + 'b,
+    >(
+        &self,
+        device_index: u8,
+        src: Read,
+        dest: Write,
+    ) -> Result<(), SpiError> {
+        self.ready_writey(
+            SpiOperation::exchange,
+            device_index,
+            Some(src),
+            Some(dest),
+        )
+    }
+
+    pub fn lock(
+        &self,
+        sender: TaskId,
+        devidx: u8,
+        cs_state: CsState,
+    ) -> Result<(), SpiError> {
+        let cs_asserted = cs_state == CsState::Asserted;
+        let devidx = usize::from(devidx);
+
+        // If we are locked there are more rules:
+        if let Some(lockstate) = &self.lock_holder.get() {
+            // The fact that we received this message _at all_ means
+            // that the sender matched our closed receive, but just
+            // in case we have a server logic bug, let's check.
+            assert!(lockstate.task == sender);
+            // The caller is not allowed to change the device index
+            // once locked.
+            if lockstate.device_index != devidx {
+                return Err(SpiError::BadDevice);
+            }
+        }
+
+        // OK! We are either (1) just locking now or (2) processing
+        // a legal state change from the same sender.
+
+        // Reject out-of-range devices.
+        let device = CONFIG.devices.get(devidx).ok_or(SpiError::BadDevice)?;
+
+        for pin in device.cs {
+            // If we're asserting CS, we want to *reset* the pin. If
+            // we're not, we want to *set* it. Because CS is active low.
+            if cs_asserted {
+                self.sys.gpio_reset(*pin);
+            } else {
+                self.sys.gpio_set(*pin);
+            }
+        }
+
+        self.lock_holder.set(Some(LockState {
+            task: sender,
+            device_index: devidx,
+        }));
+        Ok(())
+    }
+
+    pub fn release(&self, sender: TaskId) -> Result<(), SpiError> {
+        if let Some(lockstate) = &self.lock_holder.get() {
+            // The fact that we were able to receive this means we
+            // should be locked by the sender...but double check.
+            assert!(lockstate.task == sender);
+
+            let device = &CONFIG.devices[lockstate.device_index];
+
+            for pin in device.cs {
+                // Deassert CS. If it wasn't asserted, this is a no-op.
+                // If it was, this fixes that.
+                self.sys.gpio_set(*pin);
+            }
+
+            self.lock_holder.set(None);
+            Ok(())
+        } else {
+            Err(SpiError::NothingToRelease)
+        }
+    }
+
+    fn ready_writey<
+        'b,
+        BufRead: BufReader<'b>,
+        Read: Into<BufRead> + 'b,
+        BufWrite: BufWriter<'b>,
+        Write: Into<BufWrite> + 'b,
+    >(
```

Comment:
Could we lose the `Read` and `Write` generics (in favor of calling `.into()` at the call sites)? If we could, would there be any benefit (fewer `ready_writey` instantiations)?

Reply:
Good idea, this seems to work (and reduces LoC; not sure if it also produces fewer instantiations)

Done in 03a92b035972f73c952b7f8650722995290ce3d3